src/io/mod.rs

```diff
@@ -1,5 +1,113 @@
 //! Input and output of images.
+
+use crate::{error, ImageError, ImageResult};
+
 mod reader;
 pub(crate) mod free_functions;
 
 pub use self::reader::Reader;
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+/// Set of supported strict limits for a decoder
+pub struct LimitSupport {
+    /// This field indicates whether the strict limit on the number of bytes is
+    /// supported.
+    pub max_alloc_strict: bool,
+    _non_exhaustive: (),
+}
+
+impl Default for LimitSupport {
+    fn default() -> LimitSupport {
+        LimitSupport {
+            max_alloc_strict: false,
+            _non_exhaustive: (),
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+/// Resource limits for decoding.
+///
+/// Limits can be either *strict* or *non-strict*. Non-strict limits are best-effort
+/// limits where the library does not guarantee that limit wont be exceeded. Do note 
+/// that it is still considered a bug if a non-strict limit is exceeded, however as 
+/// some of the underlying decoders do . For strict
+/// limits the library makes a stronger guarantee that the limit will not be exceeded.
+/// Exceeding a strict limit is considered a critical bug. If a decoder cannot 
+/// guaranree that it will uphold a strict limit it *must* fail with 
+/// `image::error::LimitErrorKind::Unsupported`.
+///
+/// The limit check should only ever fail if a limit will be exceeded or an unsupported strict
+/// limit is used.
+pub struct Limits {
+    /// The maximum allowed image width. This limit is strict. The default is no limit.
+    pub max_image_width: Option<u32>,
+    /// The maximum allowed image height. This limit is strict. The default is no limit.
+    pub max_image_height: Option<u32>,
+    /// The maximum allowed amount of memory to be allocated by the 
+    /// decoder at any one time. This limit is non-strict by default,
+    /// set `max_alloc_strict` to make this limit strict. The default 
+    /// is 512MiB. Note that limits below 1MiB are not supported, this 
+    /// is also the case if this limit is strict.
+    pub max_alloc: Option<u64>,
```

Comment:
This is not entirely clear and it will be complicated with 'strict'. It says 'at any one time' but it omits whether this means 'calls to `alloc` specify at most size x' or 'calling decode will increased memory use by at most x'. 

The former is reasonbly easy and what decoders mostly implement at the moment—i.e. ancillary data, color tables, icc, are _not_ subtracted from the available budget. Just to be clear, I'd be in principle okay with calling all of these decoders non-strict!

However, the latter option is borderline impossible to implement and specify. Does it include memory reserved by the underlying allocator in anticipation of later use? Should it count stack usage? Is deallocated memory considered to be free again instantly? In case of an error, or an exceeded limit, should the use up to that point be report? This appears contradictory because the motivation for strictness is a guarantee in the first place but we can't specify the guarantee exactly. For strict limits there should also be an explanation that supporting them must come with 'essentially free' construction. The dilemma of `tiff` ifd-reader should not occur for very strict decoders.

The `Limits` struct is extensible so we could defer `strict` to a future iteration of limits?

Reply:
What is was thinking was something among the lines of 'accumulated calls to malloc and free (where we think of free as reserving negative memory) will never grow past `max_alloc`'. I did think about saying something about it allowing usage of '1MiB or 10%, whatever is greater' outside of the specified amount, but then the decoder might as well just remove 10% on its own.

The idea is that strict memory limits are more or less **strict** in the sense that if it is important for you that  memory usage is limited you should be able to set strict and be confident that decoder will not potentially try to allocate more memory than specified. This would of course currently lead to only a subset of decoders supporting strict limits and the docs should probably reflect that, but I was afraid of adding a table that would quickly get outdated.

Deferring `strict` to later versions could be an idea, however this would mean that `LimitSupport` would be empty for now.